futures-core/src/task/mod.rs

```diff
@@ -1,129 +1,53 @@
+pub use core::task::{UnsafeWake, Waker};
 #[cfg(feature = "std")]
-pub use self::wake::Wake;
+pub use std::task::Wake;
 
-mod context;
-pub use self::context::Context;
+pub use core::task::Context;
 
 #[cfg_attr(feature = "nightly", cfg(target_has_atomic = "ptr"))]
 mod atomic_waker;
 #[cfg_attr(feature = "nightly", cfg(target_has_atomic = "ptr"))]
 pub use self::atomic_waker::AtomicWaker;
 
-/// A custom trait object for polling tasks, roughly akin to
-/// `Box<Future<Output = ()> + Send>`.
-pub struct TaskObj {
-    ptr: *mut (),
-    poll: unsafe fn(*mut (), &mut Context) -> Poll<()>,
-    drop: unsafe fn(*mut ()),
-}
-
-impl fmt::Debug for TaskObj {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("TakObj")
-            .finish()
-    }
-}
-
-unsafe impl Send for TaskObj {}
-unsafe impl Sync for TaskObj {}
-
-/// A custom implementation of a task trait object for `TaskObj`, providing
-/// a hand-rolled vtable.
-///
-/// This custom representation is typically used only in `no_std` contexts,
-/// where the default `Box`-based implementation is not available.
-///
-/// The implementor must guarantee that it is safe to call `poll` repeatedly (in
-/// a non-concurrent fashion) with the result of `into_raw` until `drop` is
-/// called.
-pub unsafe trait UnsafePoll: Send + 'static {
-    /// Convert a owned instance into a (conceptually owned) void pointer.
-    fn into_raw(self) -> *mut ();
+pub use core::task::{TaskObj, UnsafePoll};
 
-    /// Poll the task represented by the given void pointer.
-    ///
-    /// # Safety
-    ///
-    /// The trait implementor must guarantee that it is safe to repeatedly call
-    /// `poll` with the result of `into_raw` until `drop` is called; such calls
-    /// are not, however, allowed to race with each other or with calls to `drop`.
-    unsafe fn poll(task: *mut (), cx: &mut Context) -> Poll<()>;
-
-    /// Drops the task represented by the given void pointer.
-    ///
-    /// # Safety
-    ///
-    /// The trait implementor must guarantee that it is safe to call this
-    /// function once per `into_raw` invocation; that call cannot race with
-    /// other calls to `drop` or `poll`.
-    unsafe fn drop(task: *mut ());
-}
-
-impl TaskObj {
-    /// Create a `TaskObj` from a custom trait object representation.
-    pub fn from_poll_task<T: UnsafePoll>(t: T) -> TaskObj {
-        TaskObj {
-            ptr: t.into_raw(),
-            poll: T::poll,
-            drop: T::drop,
-        }
-    }
+if_std! {
+    use std::boxed::PinBox;
 
-    /// Poll the task.
-    ///
-    /// The semantics here are identical to that for futures, but unlike
-    /// futures only an `&mut self` reference is needed here.
-    pub fn poll_task(&mut self, cx: &mut Context) -> Poll<()> {
-        unsafe {
-            (self.poll)(self.ptr, cx)
-        }
+    /// Extension trait for `TaskObj`, adding methods that require allocation.
+    pub trait TaskObjExt {
```

Comment:
What's the purpose of this trait? Since the method doesn't take `self`, it's not actually callable using method syntax, so it might as well be a free function. IMO users should either call `TaskObj::from_poll_task` directly or call the `spawn` method on the `ContextExt` trait (yet to be written).